### PR TITLE
834 new rollback

### PR DESCRIPTION
**Pull request by arkuz**

### CODE REVIEW

File: rollback1.py [hunk 1]
```diff
@@ -8,15 +8,28 @@
 
 
 class Rollback1TestCase(unittest.TestCase):
-    
+
+    def setUp(self):
+        global url, prKey, token, waitTx, host, db, pas, login
```
Comment (potehinre) [inline rollback1.py]: Use of global variables is always bad, why not use object local variables, self.url = self.conf["url"], and so on

File: rollback1.py [hunk 2]
```diff
@@ -26,96 +39,171 @@ def create_contract(self, data):
             dataC = data
         res = self.call("NewContract", dataC)
         return name, code
-        
+
+    def addNotification(self):
+        # create contract, wich added record in notifications table
+        function = """DBInsert("notifications", "recipient->member_id,notification->type,notification->header,notification->body", "-8399130570195839739", 1, "Message header", "Message body")"""
+        body = """ { \n data {}	\n conditions {} \n	action { \n """ + function + """ \n } }"""
```
Comment (potehinre) [inline rollback1.py]: better way:
"""
{ 
data{}
conditions{}
action {
%s
}""" % (function,)
much more readable

File: rollback1.py [hunk 3]
```diff
@@ -26,96 +39,171 @@ def create_contract(self, data):
...
+        res = self.call("EditTable", dataEdit)
+        # call contract, wich added record in notification table
+        res = self.call(name, "")
+        # change permission for notifications table back
+        permission = """{"insert": "ContractAccess(\\"Notifications_Single_Send_map\\",\\"Notifications_Roles_Send_map\\")", "update": "ContractConditions(\\"MainCondition\\")", "new_column": "ContractConditions(\\"MainCondition\\")"}"""
+        dataEdit = {"Name": "notifications", "Permissions": permission}
+        res = self.call("EditTable", dataEdit)
+
+    def getCountTable(self,name):
+        return utils.getCountTable(host, db, login, pas, name)
```
Comment (potehinre) [inline rollback1.py]: maybe its better just pass db config to functions like that, so helper function getCountTable is useless
utils.getCountTable(self.conf)

File: rollback1.py [hunk 4]
```diff
@@ -8,15 +8,28 @@
 
 
 class Rollback1TestCase(unittest.TestCase):
-    
+
+    def setUp(self):
+        global url, prKey, token, waitTx, host, db, pas, login
+        self.conf = config.readMainConfig()
+        url = self.conf["url"]
```
Comment (potehinre) [inline rollback1.py]: what the point of all that variables, i think it's better just readMainConfig(), and use self.conf everywhere

File: rollback1.py [hunk 5]
```diff
@@ -26,96 +39,171 @@ def create_contract(self, data):
...
+        binCount = self.getCountTable("1_binaries")
+        name = "image_" + utils.generate_random_name()
+        path = os.path.join(os.getcwd(), "fixtures", "image2.jpg")
+        with open(path, 'rb') as f:
+            file = f.read()
+        files = {'Data': file}
+        data = {"Name": name, "ApplicationId": 1}
+        resp = utils.call_contract_with_files(url, prKey, "UploadBinary", data,
+                                              files, token)
+        while True:
```
Comment (potehinre) [inline rollback1.py]: what if this record will never be added to the table, the test will stuck, so i recomend add test fail after some timeout, maybe 30s?

File: rollback1.py [hunk 6]
```diff
@@ -26,96 +39,171 @@ def create_contract(self, data):
...
+        data = {"Name": tableName,
+                "Columns": column, "ApplicationId": 1,
+                "Permissions": permission}
+        res = self.call("NewTable", data)
+        return tableName
+
+    def insertToUserTable(self, tableName):
+        # create contarct, wich added record in created table
+        function = "DBInsert(\"" + tableName + "\", \"MyName\" , \"insert\" )"
+        body = """ { \n data {}	\n conditions {} \n	action { \n """ + function + """ \n } }"""
```
Comment (potehinre) [inline rollback1.py]: same stuff as before

File: rollback1.py [hunk 7]
```diff
@@ -26,96 +39,171 @@ def create_contract(self, data):
...
+        code, name = utils.generate_name_and_code(body)
+        data = {"Value": code, "ApplicationId": 1,
+                "Conditions": "true"}
+        res = self.call("NewContract", data)
+        # call contarct, wich added record in created table
+        res = self.call(name, data)
+
+    def updateUserTable(self, tableName):
+        # create contarct, wich updated record in created table
+        function = "DBUpdate(\"" + tableName + "\", 1, \"MyName,ver_on_null\" ,\"update\",\"update\" )"
```
Comment (potehinre) [inline rollback1.py]: same stuff as before

File: rollback1.py [hunk 8]
```diff
@@ -183,29 +271,39 @@ def edit_sign(self, name):
...
-        db = self.conf["dbName"]
-        login = self.conf["login"]
-        pas = self.conf["pass"]
-        waitTx = self.conf["time_wait_tx_in_block"]
+        self.addNotification()
+        self.addBinary()
+        tableName = self.addUserTable()
+        self.insertToUserTable(tableName)
+        # Save to file block id for rollback
+        RollbackBlockId = funcs.get_max_block_id(url, token)
```
Comment (potehinre) [inline rollback1.py]: not pythonic name, rollback_block_id

File: utils.py [hunk 9]
```diff
@@ -226,6 +226,29 @@ def getCountDBObjects(dbHost, dbName, login, password):
 		tablesCount[table[2:]] = getCountTable(dbHost, dbName, login, password, table)
 	return tablesCount
 
+def getTableColumnNames(dbHost, dbName, login, password, table):
+	connect = psycopg2.connect(host=dbHost, dbname=dbName, user=login, password=password)
```
Comment (potehinre) [inline utils.py]: what the point connect every time, it's better connect once at begging of the test, at setUp functions, then pass connection to helper functions like that

File: utils.py [hunk 10]
```diff
@@ -226,6 +226,29 @@ def getCountDBObjects(dbHost, dbName, login, password):
...
+	cursor = connect.cursor()
+	query = "SELECT pg_attribute.attname FROM pg_attribute, pg_class WHERE pg_class.relname='"+\
+			table+"' AND pg_class.relfilenode=pg_attribute.attrelid AND pg_attribute.attnum>0"
+	cursor.execute(query)
+	col = {}
+	col = cursor.fetchall()
+	return col
+
+def getUserTableState(dbHost, dbName, login, password, userTable):
+	connect = psycopg2.connect(host=dbHost, dbname=dbName, user=login, password=password)
```
Comment (potehinre) [inline utils.py]: Same stuff

File: utils.py [hunk 11]
```diff
@@ -226,6 +226,29 @@ def getCountDBObjects(dbHost, dbName, login, password):
...
+	cursor = connect.cursor()
+	cursor.execute("SELECT * FROM \"" + userTable + "\"")
+	res = cursor.fetchall()
+	col = getTableColumnNames(dbHost, dbName, login, password, userTable)
+	table = {}
+	i = 0
+	while i<len(col):
```
Comment (potehinre) [inline utils.py]: use for cycle for collection iteration like that, 
for i in range(len(col)):

File: rollback1.py [hunk 12]
```diff
@@ -183,29 +271,39 @@ def edit_sign(self, name):
...
+        file = os.path.join(os.getcwd(), "blockId.txt")
+        with open(file, 'w') as f:
+            f.write(str(RollbackBlockId))
+        # Save to file user table name
+        tableNameWithPrefix = "1_" + tableName
```
Comment (potehinre) [inline rollback1.py]: same

File: rollback1.py [hunk 13]
```diff
@@ -183,29 +271,39 @@ def edit_sign(self, name):
...
+        file = os.path.join(os.getcwd(), "userTableName.txt")
+        with open(file, 'w') as f:
+            f.write(str(tableNameWithPrefix))
+        # Save to file user table state
+        dbUserTableInfo = utils.getUserTableState(host, db, login, pas, tableNameWithPrefix)
```
Comment (potehinre) [inline rollback1.py]: same

File: rollback1.py (hunk 1 above)
Comment (arkuz) [reply rollback1.py]: I can't change that now. Such changes are needed in all tests.

File: rollback1.py (hunk 2 above)
Comment (arkuz) [reply rollback1.py]: fixed

File: rollback1.py (hunk 5 above)
Comment (arkuz) [reply rollback1.py]: Fixed

File: rollback1.py (hunk 6 above)
Comment (arkuz) [reply rollback1.py]: fixed

File: rollback1.py (hunk 7 above)
Comment (arkuz) [reply rollback1.py]: fixed

File: rollback1.py (hunk 8 above)
Comment (arkuz) [reply rollback1.py]: fixed

File: rollback1.py (hunk 12 above)
Comment (arkuz) [reply rollback1.py]: fixed

File: rollback1.py (hunk 13 above)
Comment (arkuz) [reply rollback1.py]: used camelCase similar to the already existing variables

File: utils.py (hunk 9 above)
Comment (arkuz) [reply utils.py]: Made similar to already existing funcs

File: utils.py (hunk 10 above)
Comment (arkuz) [reply utils.py]: Made similar to already existing funcs

**Pull request merged**